### PR TITLE
Fix Scaladoc crash when extending non-Scala-3 classes (long-term fix for 3.4.0)

### LINKED ISSUE #15927: Scaladoc generation fails when extending some Java-defined or Scala-2-defined classes
## Compiler version

3.2.0.rc4 and sbt 1.7.1,

If you're not sure what version you're using, run `print scalaVersion` from sbt
(if you're running scalac manually, use `scalac -version` instead).

## Minimized code

```
sbt clean compile doc package
```

```scala
//the following is **BugTest.scala**


//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** @author  Gitzinger
 *
 *  @title   `BugTest` for Handle Dates and Times
 */

package test4education



//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** The `BugTest`  class 
 */
abstract class BugTest () extends Numeric [BugTest] with Serializable:

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /**  a `BugTest` 
     */
    def what() = println("a bug")
  
    

end BugTest

 ```

## Output (click arrow to expand)
<details>

```scala
// TODO add output here
[info] Main Scala API documentation to ../.../scala-3.2.0-RC4/api...
Problem parsing src/main/scala/BugTest.scala:[159..167..529], documentation may not be generated.
java.lang.AssertionError: assertion failed
  | => rat scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
	at dotty.tools.dotc.util.Spans$Span$.start$extension(Spans.scala:45)
	at dotty.tools.dotc.util.SourcePosition.start(SourcePosition.scala:52)
	at scala.quoted.runtime.impl.QuotesImpl$reflect$PositionMethods$.start(QuotesImpl.scala:2813)
	at scala.quoted.runtime.impl.QuotesImpl$reflect$PositionMethods$.start(QuotesImpl.scala:2813)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.getParentsAsTreeSymbolTuples$$anonfun$1(ClassLikeSupport.scala:260)
	at scala.collection.Iterator$$anon$6.hasNext(Iterator.scala:472)
	at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:576)
	at scala.collection.immutable.List.prependedAll(List.scala:152)
	at scala.collection.immutable.List$.from(List.scala:684)
	at scala.collection.immutable.List$.from(List.scala:681)
	at scala.collection.IterableOps$WithFilter.map(Iterable.scala:891)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.getParentsAsTreeSymbolTuples(ClassLikeSupport.scala:264)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.getParentsAsTreeSymbolTuples$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.getParentsAsTreeSymbolTuples(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.getParentsAsLinkToTypes(ClassLikeSupport.scala:253)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.getParentsAsLinkToTypes$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.getParentsAsLinkToTypes(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.mkClass(ClassLikeSupport.scala:106)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.mkClass$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.mkClass(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.parseClasslike(ClassLikeSupport.scala:290)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.parseClasslike$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.parseClasslike(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.parseInheritedMember$$anonfun$1(ClassLikeSupport.scala:190)
	at dotty.tools.scaladoc.tasty.TastyParser.processTreeOpt(TastyParser.scala:204)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.parseInheritedMember(ClassLikeSupport.scala:202)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.extractMembers$$anonfun$2(ClassLikeSupport.scala:221)
	at scala.collection.immutable.List.flatMap(List.scala:293)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.extractMembers(ClassLikeSupport.scala:221)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.extractMembers$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.extractMembers(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.extractPatchedMembers(ClassLikeSupport.scala:226)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.extractPatchedMembers$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.extractPatchedMembers(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.mkClass(ClassLikeSupport.scala:113)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.mkClass$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.mkClass(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.parseClasslike(ClassLikeSupport.scala:290)
	at dotty.tools.scaladoc.tasty.ClassLikeSupport.parseClasslike$(ClassLikeSupport.scala:15)
	at dotty.tools.scaladoc.tasty.TastyParser.parseClasslike(TastyParser.scala:169)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.traverseTree(TastyParser.scala:223)
	at scala.quoted.Quotes$reflectModule$TreeTraverser.foldTree(Quotes.scala:4666)
	at scala.quoted.Quotes$reflectModule$TreeTraverser.foldTree$(Quotes.scala:4662)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.foldTree(TastyParser.scala:211)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.foldTree(TastyParser.scala:211)
	at scala.quoted.Quotes$reflectModule$TreeAccumulator.foldTrees$$anonfun$1(Quotes.scala:4554)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at scala.quoted.Quotes$reflectModule$TreeAccumulator.foldTrees(Quotes.scala:4554)
	at scala.quoted.Quotes$reflectModule$TreeAccumulator.foldTrees$(Quotes.scala:4549)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.foldTrees(TastyParser.scala:211)
	at scala.quoted.Quotes$reflectModule$TreeAccumulator.foldOverTree(Quotes.scala:4617)
	at scala.quoted.Quotes$reflectModule$TreeAccumulator.foldOverTree$(Quotes.scala:4549)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.foldOverTree(TastyParser.scala:211)
	at scala.quoted.Quotes$reflectModule$TreeTraverser.traverseTreeChildren(Quotes.scala:4668)
	at scala.quoted.Quotes$reflectModule$TreeTraverser.traverseTreeChildren$(Quotes.scala:4662)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.traverseTreeChildren(TastyParser.scala:211)
	at scala.quoted.Quotes$reflectModule$TreeTraverser.traverseTree(Quotes.scala:4664)
	at scala.quoted.Quotes$reflectModule$TreeTraverser.traverseTree$(Quotes.scala:4662)
	at dotty.tools.scaladoc.tasty.TastyParser$Traverser$2$.traverseTree(TastyParser.scala:219)
	at dotty.tools.scaladoc.tasty.TastyParser.parseRootTree(TastyParser.scala:228)
	at dotty.tools.scaladoc.tasty.ScaladocTastyInspector.$anonfun$4(TastyParser.scala:123)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at dotty.tools.scaladoc.tasty.ScaladocTastyInspector.postProcess$$anonfun$2(TastyParser.scala:42)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.scaladoc.tasty.ScaladocTastyInspector.postProcess(TastyParser.scala:42)
	at scala.tasty.inspector.OldTastyInspector$TastyInspectorFinishPhase$1.runOn(OldTastyInspector.scala:91)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:234)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:245)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:253)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:262)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:262)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:192)
	at dotty.tools.dotc.fromtasty.TASTYRun.compile(TASTYRun.scala:14)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
	at dotty.tools.dotc.Driver.process(Driver.scala:195)
	at scala.tasty.inspector.OldTastyInspector.inspectFilesInContext(OldTastyInspector.scala:72)
	at scala.tasty.inspector.OldTastyInspector.inspectFilesInContext$(OldTastyInspector.scala:22)
	at scala.tasty.inspector.DocTastyInspector.inspectFilesInContext(DocTastyInspector.scala:5)
	at dotty.tools.scaladoc.tasty.ScaladocTastyInspector.result(TastyParser.scala:147)
	at dotty.tools.scaladoc.ScalaModuleProvider$.mkModule(ScalaModuleProvider.scala:11)
	at dotty.tools.scaladoc.Scaladoc$.run(Scaladoc.scala:230)
	at dotty.tools.scaladoc.Scaladoc$.run$$anonfun$1(Scaladoc.scala:72)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.Option.map(Option.scala:242)
	at dotty.tools.scaladoc.Scaladoc$.run(Scaladoc.scala:76)
	at dotty.tools.dottydoc.Main$.process(Main.scala:25)
	at dotty.tools.dottydoc.Main.process(Main.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at xsbt.DottydocRunner.run(DottydocRunner.java:61)
	at xsbt.ScaladocInterface.run(ScaladocInterface.java:11)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at sbt.internal.inc.AnalyzingCompiler.invoke(AnalyzingCompiler.scala:329)
	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:175)
	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:133)
	at sbt.Doc$.$anonfun$scaladoc$1(Doc.scala:52)
	at sbt.Doc$.$anonfun$scaladoc$1$adapted(Doc.scala:40)
	at sbt.RawCompileLike$.$anonfun$prepare$1(RawCompileLike.scala:79)
	at sbt.RawCompileLike$.$anonfun$prepare$1$adapted(RawCompileLike.scala:72)
	at sbt.RawCompileLike$.$anonfun$cached$4(RawCompileLike.scala:63)
	at sbt.RawCompileLike$.$anonfun$cached$4$adapted(RawCompileLike.scala:61)
	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
	at sbt.RawCompileLike$.$anonfun$cached$1(RawCompileLike.scala:68)
	at sbt.RawCompileLike$.$anonfun$cached$1$adapted(RawCompileLike.scala:52)
	at sbt.Defaults$.$anonfun$docTaskSettings$4(Defaults.scala:2157)
	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
	at sbt.Execute.work(Execute.scala:291)
	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
[warn] Companion for class xxx exists but is missing in classlike map


```
</details>

### PR DESCRIPTION
**Pull request by jan-pieter**
Extend the PositionMethods with an exists method, since the other methods are not safe to call on non-existing Spans.

Not eligible for backport to 3.3 because of API addition. See #16759 for a workaround that uses the current API.

Fixes #15927

### CODE REVIEW

**Comment by jan-pieter** [comment]
As expected, MiMa fails for this change.

**Comment by SethTisue** [comment]
(marking as draft for now since it's #16759 that should be merged first)

**Comment by SethTisue** [comment]
> As expected, MiMa fails for this change

I believe we need to do something about that, but I'm not clear on what the right thing to do is.

If we believe the change is safe, then we can mollify CI by adding it an exclusion to `project/MiMaFilters.scala`.

But I think the change isn't safe, because adding a method to a trait isn't backwards binary compatible unless you include a default implementation. And it matters because `Quotes` is part of the standard library, it's not compiler internals where we're free to break stuff.

Can we just `def exists: Boolean = false` (or `= true`? 🤔), so a default implementation is there to satisfy MiMa, and be done with it?

In practice, I don't think there is any realistic scenario in which the default method — or the abstract method, if we leave it abstract — would end up being called. The only call to `exists` that exists is the one we ourselves are adding to `ClassLikeSupport.scala`. In order for that call to end up being routed through the abstract or default at runtime, someone would have to be using `ClassLikeSupport` directly rather than in the context of the Scaladoc tool, but `ClassLikeSupport` is itself in an `impl` package that is considered internal.

In short, I think it would be safe to simply add a MiMa exclusion. I also think it would be safe to have `def exists: Boolean = false` (and this is probably marginally preferable to defaulting to `true`, since `false` doesn't promise anything to the caller) — but there is perhaps a slight danger of providing a default implementation, which is that it makes it possible to forget to override it, perhaps in future refactoring of the codebase.

In short, I suggest we add the MiMa exclusion and be done with it.

@dwijnand Do you agree with this analysis? I've written the above without hedging, as if I were 100% sure, but I'm not really that sure my reasoning is correct and complete.

**Comment by dwijnand** [comment]
> But I think the change isn't safe, because adding a method to a trait isn't backwards binary compatible unless you include a default implementation. And it matters because `Quotes` is part of the standard library, it's not compiler internals where we're free to break stuff.

`QuotesImpl` is the only implementation of the interface, so we don't need to worry about defaulting it.

Also, we don't need to worry about forwards compatibility across minor versions because we are free to break it: code that was compiled against 3.4.0 (and used Position#exists) cannot be consumed by code being compiled with 3.3.0 because of tasty versions.

**Comment by julienrf** [comment]
> Also, we don't need to worry about forwards compatibility across minor versions because we are free to break it: code that was compiled against 3.4.0 (and used Position#exists) cannot be consumed by code being compiled with 3.3.0 because of tasty versions.

Adding an abstract method to a trait breaks the compatibility in the backwards direction: existing code like `new TypeThatNowHasAnAbstractMethod()` will fail at run-time because the new version of that class can’t be instantiated.

But I agree with your point that we should probably not worry too much about the `Quotes` API because it unlikely that there are other implementations of it… We could still add a point about that in the release notes, though.

**Comment by dwijnand** [comment]
> Adding an abstract method to a trait breaks the compatibility in the backwards direction: existing code like `new TypeThatNowHasAnAbstractMethod()` will fail at run-time because the new version of that class can’t be instantiated.

You're right.

> But I agree with your point that we should probably not worry too much about the `Quotes` API because it is unlikely that there are other implementations of it… We could still add a point about that in the release notes, though.

Don't even think it's worth release noting.  Perhaps it's worth a sentence in the Quotes doc, as in "not intended to be implemented inherited/implemented".  It also has a self type with traits that are in scala.quotes.runtime which, if we use scala.runtime as well as their docs, implies it's not public API, making it impossible to legally implement Quotes.

**Comment by Sporarum** [comment]
We had a similar issue with position when working on https://github.com/lampepfl/dotty/pull/14810,

Since this is the long term solution, I encourage you to also refactor
https://github.com/lampepfl/dotty/blob/main/scaladoc/src/dotty/tools/scaladoc/tasty/SymOps.scala#L157-L173
And other snippets going around the previous limitations
(having the wrong way to do something in the code might make people simply copy it without looking for the right way)

This might also be the place to solve https://github.com/lampepfl/dotty/issues/16619, for example by adding an assert to `exists` ?

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 1]
```diff
@@ -2850,6 +2850,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
 
     given PositionMethods: PositionMethods with
       extension (self: Position)
+        def exists: Boolean = self.exists
```
Comment (Sporarum) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2853]: Could these be an extension export ?

**Comment by dwijnand** [comment]
What's the issue with accepting that some positions don't exist?  I'm all with you on your ideal, but the reality seems like trees with non-existing positions crop up and either there's some support for identifying them or people are stuck doing hack arounds or making their users suffer crashes.

Seems like an uphill battle and not a hill to die on - to me.

**Comment by nicolasstucki** [comment]
The missing position bug was already fixed.

We follow up with #19250, where we make the the API resilient to missing positions. There is no need to have `exists` method.
